File: scripts/prerequisites.py
#!/usr/bin/env python3
"""
prerequisites.py

PURPOSE: Unified Phase 1 setup - Verify environment for any Prismatic workflow.

Supports different check levels based on --type:
  - component: Prism CLI + auth only
  - integration: Prism CLI + auth only

USAGE:
  python prerequisites.py <name> --type component
  python prerequisites.py <name> --type integration

EXIT CODES:
  0 - Success: All prerequisites verified
  1 - Error: Invalid usage or prerequisites not met
  2 - Error: Prismatic not authenticated
"""

import argparse
import json
import os
import re
import subprocess
import sys

# Add shared directory to path
SHARED_DIR = os.path.join(os.path.dirname(__file__), "shared")
sys.path.insert(0, SHARED_DIR)

from project_directory import (
    ensure_session_directory,
    get_project_root,
)
from timing import print_timing_summary, timed_step


# Session type mapping
SESSION_TYPE_MAP = {
    "component": "components",
    "integration": "integrations",
}


def print_section(title):
    """Print a section header."""
    print("")
    print("=" * 60)
    print(f"  {title}")
    print("=" * 60)
    print("")


def validate_name(name):
    """Validate name follows conventions."""
    pattern = r"^[a-z][a-z0-9-]*[a-z0-9]$|^[a-z]$"
    return bool(re.match(pattern, name))


@timed_step("Check Prism CLI")
def check_prism_installed():
    """Check if Prism CLI is installed."""
    try:
        result = subprocess.run(
            ["prism", "--version"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode == 0:
            version = result.stdout.strip()
            print(f"Prism CLI installed: {version}")
            return version
    except FileNotFoundError:
        pass
    except subprocess.TimeoutExpired:
        print("Prism CLI check timed out")
    return None


@timed_step("Install Prism CLI")
def install_prism():
    """Install Prism CLI via npm."""
    print("Installing Prism CLI...")
    try:
        result = subprocess.run(
            ["npm", "install", "-g", "--no-audit", "--no-fund",
             "--no-update-notifier", "@prismatic-io/prism"],
            capture_output=True, text=True, timeout=120,
        )
        if result.returncode == 0:
            verify = subprocess.run(
                ["prism", "--version"],
                capture_output=True, text=True, timeout=10,
            )
            if verify.returncode == 0:
                version = verify.stdout.strip()
                print(f"Prism CLI installed: {version}")
                return True
        print("Installation failed")
        if result.stderr:
            for line in result.stderr.strip().split("\n")[:5]:
                print(f"   {line}")
        return False
    except FileNotFoundError:
        print("npm not found - Node.js must be installed first")
        return False
    except subprocess.TimeoutExpired:
        print("Installation timed out (2 minutes)")
        return False


@timed_step("Verify Authentication")
def check_logged_in():
    """Check if user is logged in to Prism."""
    try:
        result = subprocess.run(
            ["prism", "me"],
            capture_output=True, text=True, timeout=30,
        )
        if result.returncode == 0:
            output = result.stdout.strip()
            user_info = {}
            for line in output.split("\n"):
                if ":" in line:
                    key, value = line.split(":", 1)
                    user_info[key.strip().lower()] = value.strip()

            name = user_info.get("name", "Unknown")
            email = user_info.get("email", "Unknown")
            endpoint = user_info.get("endpoint url", user_info.get("endpoint", "Unknown"))

            print(f"Logged in as: {email}")
            print(f"   Name: {name}")
            print(f"   Endpoint: {endpoint}")
            return {"name": name, "email": email, "endpoint": endpoint}
        return None
    except FileNotFoundError:
        print("Prism CLI not found - install with: npm install -g @prismatic-io/prism")
        return None
    except subprocess.TimeoutExpired:
        print("Authentication check timed out")
        return None
    except Exception as e:
        print(f"Error checking authentication: {e}")
        return None


def main():
    parser = argparse.ArgumentParser(
        description="Unified prerequisites check for Prismatic workflows"
    )
    parser.add_argument("name", help="Name for the session (e.g., 'canny', 'hubspot-crm')")
    parser.add_argument(
        "--type", choices=["component", "integration"],
        default="component", help="Type of workflow (default: component)"
    )
    args = parser.parse_args()

    workflow_type = args.type
    name = args.name

    type_labels = {
        "component": "Prismatic Component Builder",
        "integration": "Prismatic CNI Builder",
    }

    print(f"{type_labels[workflow_type]} - Phase 1 Setup")
    print("")

    if not validate_name(name):
        print("Invalid name")
        print("   Name must be lowercase with hyphens (e.g., 'canny', 'hubspot-crm')")
        return 1

    all_passed = True
    user_info = None

    # Common checks: Prism CLI + auth
    print_section("Checking Prism CLI")
    version = check_prism_installed()

    if not version:
        response = input("Prism CLI not found. Install now? [Y/n]: ").strip().lower()
        if response in ("", "y", "yes"):
            if not install_prism():
                all_passed = False
        else:
            print("\nTo install: npm install -g @prismatic-io/prism")
            all_passed = False

    if all_passed or version:
        print_section("Verifying Authentication")
        user_info = check_logged_in()
        if not user_info:
            print("\nNot logged in to Prismatic. Please run: prism login")
            all_passed = False

    # Create session directory
    session_type = SESSION_TYPE_MAP[workflow_type]
    session_dir = None
    if all_passed:
        session_dir = ensure_session_directory(name, session_type)
        print(f"\nSession directory created: {session_dir}")

    # Final status
    print("")
    print("=" * 60)
    if all_passed:
        print("  PHASE 1 SETUP COMPLETE")
    else:
        print("  PHASE 1 SETUP INCOMPLETE")
    print("=" * 60)

    print_timing_summary()
    print("")

    if all_passed:
        requirements_file = os.path.join(session_dir, "requirements.json")

        output = {
            "status": "ready",
            "name": name,
            "type": workflow_type,
            "session_dir": session_dir,
            "requirements_file": requirements_file,
            "user": user_info,
        }

        print("Ready for Phase 2 - Requirements Gathering!")
        print("")
        print(f"Name: {name}")
        print(f"Type: {workflow_type}")
        print(f"Session directory: {session_dir}")
        if user_info:
            print(f"Prismatic: {user_info.get('endpoint', 'Unknown')}")
            print(f"User: {user_info.get('email', 'Unknown')}")
        print("")
        print("Next: Run requirements gathering")
        print(f"   python scripts/gather_requirements.py \\")
        print(f"     scripts/questions/{workflow_type}.json \\")
        print(f"     {requirements_file}")
        print("")
        print("--- Setup Data (JSON) ---")
        print(json.dumps(output, indent=2))

    if not all_passed:
        return 1 if not user_info else 2

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugin/scripts/prerequisites.ts
#!/usr/bin/env npx tsx
import { spawnSync } from "node:child_process";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { type CliConfig, parseCliArgs } from "./shared/cli-help.js";
import { confineToProjectRoot, ensureSessionDirectory } from "./shared/project-directory.js";
import { printTimingSummary, timedStep } from "./shared/timing.js";

const CLI = {
  command: "prismatic-tools prerequisites",
  description: "Prepare a component or integration workflow session.",
  positionals: [{ name: "name", required: true }],
  options: [
    {
      name: "type",
      type: "string",
      value: "component|integration",
      required: true,
      choices: ["component", "integration"],
      description: "Workflow type.",
    },
    { name: "existing", type: "string", value: "dir", description: "Existing project directory." },
  ],
} as const satisfies CliConfig;

const SESSION_TYPE_MAP: Record<string, string> = {
  component: "components",
  integration: "integrations",
};

function printSection(title: string): void {
  console.log("");
  console.log("=".repeat(60));
  console.log(`  ${title}`);
  console.log("=".repeat(60));
  console.log("");
}

function validateName(name: string): boolean {
  return /^[a-z][a-z0-9-]*[a-z0-9]$|^[a-z]$/.test(name);
}

function checkPrismInstalled(): string | null {
  return timedStep("Check Prism CLI", () => {
    try {
      const result = spawnSync("prism", ["--version"], {
        encoding: "utf-8",
        timeout: 10000,
      });
      if (result.status === 0) {
        const version = result.stdout.trim();
        console.log(`Prism CLI installed: ${version}`);
        return version;
      }
    } catch {
      /* not installed */
    }
    return null;
  });
}

function checkLoggedIn(): Record<string, string> | null {
  return timedStep("Verify Authentication", () => {
    try {
      const result = spawnSync("prism", ["me"], {
        encoding: "utf-8",
        timeout: 30000,
      });
      if (result.status === 0) {
        const userInfo: Record<string, string> = {};
        for (const line of result.stdout.trim().split("\n")) {
          const colonIdx = line.indexOf(":");
          if (colonIdx > -1) {
            const key = line.slice(0, colonIdx).trim().toLowerCase();
            const value = line.slice(colonIdx + 1).trim();
            userInfo[key] = value;
          }
        }
        const name = userInfo.name ?? "Unknown";
        const email = userInfo.email ?? "Unknown";
        const endpoint = userInfo["endpoint url"] ?? userInfo.endpoint ?? "Unknown";
        console.log(`Logged in as: ${email}`);
        console.log(`   Name: ${name}`);
        console.log(`   Endpoint: ${endpoint}`);
        return { name, email, endpoint };
      }
      return null;
    } catch {
      return null;
    }
  });
}

function main(): number {
  const { values, positionals } = parseCliArgs(process.argv.slice(2), CLI);
  const name = positionals.at(-1) ?? "";
  const workflowType = values.type;
  const existingDir = typeof values.existing === "string" ? values.existing : "";

  const typeLabels: Record<string, string> = {
    component: "Prismatic Component Builder",
    integration: "Prismatic CNI Builder",
  };

  console.log(`${typeLabels[workflowType]} - Phase 1 Setup`);
  console.log("");

  if (!validateName(name)) {
    console.log("Invalid name");
    console.log("   Name must be lowercase with hyphens (e.g., 'canny', 'hubspot-crm')");
    return 1;
  }

  // Validate --existing project if provided
  let existingProject = "";
  if (existingDir) {
    let resolvedDir: string;
    try {
      resolvedDir = confineToProjectRoot(existingDir);
    } catch (e) {
      console.log(`\n${(e as Error).message}`);
      return 1;
    }
    existingProject = resolvedDir;
    const pkgPath = join(resolvedDir, "package.json");
    const indexPath = join(resolvedDir, "src/index.ts");

    if (!existsSync(pkgPath)) {
      console.log(`\nNot a valid project: missing package.json in ${resolvedDir}`);
      return 1;
    }

    try {
      const pkg = JSON.parse(readFileSync(pkgPath, "utf-8"));
      const deps = { ...pkg.dependencies, ...pkg.devDependencies };
      if (!("@prismatic-io/spectral" in deps)) {
        console.log(`\nNot a Prismatic project: @prismatic-io/spectral not in dependencies`);
        return 1;
      }
    } catch {
      console.log(`\nFailed to read package.json in ${resolvedDir}`);
      return 1;
    }

    if (!existsSync(indexPath)) {
      console.log(`\nNot a valid CNI project: missing src/index.ts`);
      return 1;
    }

    console.log(`Existing project validated: ${resolvedDir}`);
  }

  let allPassed = true;
  let userInfo: Record<string, string> | null = null;

  // Check Prism CLI
  printSection("Checking Prism CLI");
  const version = checkPrismInstalled();

  if (!version) {
    // Can't prompt interactively in agent context — just fail
    console.log("\nPrism CLI not found. Install with: npm install -g @prismatic-io/prism");
    allPassed = false;
  }

  if (allPassed || version) {
    printSection("Verifying Authentication");
    userInfo = checkLoggedIn();
    if (!userInfo) {
      console.log("\nNot logged in to Prismatic. Please run: prism login");
      allPassed = false;
    }
  }

  // Create session directory
  const sessionType = SESSION_TYPE_MAP[workflowType];
  let sessionDir: string | null = null;
  if (allPassed) {
    sessionDir = ensureSessionDirectory(name, sessionType);
    console.log(`\nSession directory created: ${sessionDir}`);
  }

  // Final status
  console.log("");
  console.log("=".repeat(60));
  console.log(allPassed ? "  PHASE 1 SETUP COMPLETE" : "  PHASE 1 SETUP INCOMPLETE");
  console.log("=".repeat(60));

  printTimingSummary();
  console.log("");

  if (allPassed && sessionDir) {
    const requirementsFile = `${sessionDir}/requirements.json`;
    if (!existsSync(requirementsFile)) {
      writeFileSync(requirementsFile, JSON.stringify({}, null, 2));
    }
    const output: Record<string, unknown> = {
      status: "ready",
      name,
      type: workflowType,
      session_dir: sessionDir,
      requirements_file: requirementsFile,
      user: userInfo,
    };

    if (existingProject) {
      output.existing_project = existingProject;
    }

    console.log("Ready for Phase 2 - Requirements Gathering!");
    console.log("");
    console.log(`Name: ${name}`);
    console.log(`Type: ${workflowType}`);
    console.log(`Session directory: ${sessionDir}`);
    if (userInfo) {
      console.log(`Prismatic: ${userInfo.endpoint ?? "Unknown"}`);
      console.log(`User: ${userInfo.email ?? "Unknown"}`);
    }
    console.log("");
    console.log("Next: Run requirements gathering");
    console.log(
      `   prismatic-tools update-tasks --session ${name} --type ${workflowType} --actionable`,
    );
    console.log("");
    console.log("--- Setup Data (JSON) ---");
    console.log(JSON.stringify(output, null, 2));
  }

  if (!allPassed) {
    return userInfo ? 2 : 1;
  }

  return 0;
}

process.exit(main());
